File: leetcode_java/BT_PreorderTraversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> list = new ArrayList<Integer>();
        if (root == null) return list;
        
        Stack<TreeNode> s = new Stack<TreeNode>();
        s.push(root);
        
        while (! s.empty()) {
            TreeNode n = s.peek();
            s.pop();
            
            list.add(n.val);
            
            if (n.right != null) s.push(n.right);
            if (n.left != null) s.push(n.left);
        }
        
        return list;
    }
}

/**
Binary Tree Preorder Traversal
Difficulty: Medium

Given a binary tree, return the preorder traversal of its nodes' values.

For example:
Given binary tree {1,#,2,3},

   1
    \
     2
    /
   3

return [1,2,3].

Note: Recursive solution is trivial, could you do it iteratively?
 */

File: leetcode/BT_PreorderTraversal.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution2 {
public:
    vector<int> preorderTraversal(TreeNode *root) {
        // IMPORTANT: Please reset any member data you declared, as
        // the same Solution instance will be reused for each test case.
        vector<int> ans;
        if (! root) return ans;
        
        stack<TreeNode *> s;
        s.push(root);
        
        while (! s.empty()) {
            TreeNode * n = s.top();
            s.pop();
            ans.push_back(n->val);
            
            if (n->right) s.push(n->right);
            if (n->left) s.push(n->left);
        }
        
        return ans;
    }
};

class Solution {
Public:
   vector<int> preorderTraversal2(TreeNode* root) {
       vector<int> result;
       pre(root, result);
       return result;
   }

   void pre(TreeNode* root, vector<int>& result) {
       if (! root) return;
       result.push_back(root->val);
       pre(root->left, result);
       pre(root->right, result);
   }
};
